File: frontend/src/api/receivables.ts
import { api } from './client'
import type { Receivable } from '@/types'

export interface ReceivableListParams {
  customer_id?: string
  status?: string
}

export interface BulkPayResult {
  total_paid: number
  receivables_updated: number
  customer_balance_due: number
}

export async function listReceivables(params?: ReceivableListParams): Promise<Receivable[]> {
  const { data } = await api.get<Receivable[]>('/receivables', { params })
  return data
}

export async function registerPayment(
  id: string,
  amount: number,
  paid_by_name?: string
): Promise<Receivable> {
  const { data } = await api.post<Receivable>(`/receivables/${id}/pay`, {
    amount,
    paid_by_name,
  })
  return data
}

export async function bulkPayReceivables(
  customer_id: string,
  amount: number
): Promise<BulkPayResult> {
  const { data } = await api.post<BulkPayResult>('/receivables/bulk-pay', {
    customer_id,
    amount,
  })
  return data
}

File: backend/app/api/v1/receivables.py
import uuid

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models.receivable import ReceivableStatus
from app.models.user import User
from app.schemas.receivable import PaymentCreate, ReceivableOut
from app.services import receivable_service

router = APIRouter(prefix="/receivables", tags=["receivables"])


@router.get("", response_model=list[ReceivableOut])
async def list_receivables(
    customer_id: uuid.UUID | None = None,
    status: ReceivableStatus | None = None,
    open_only: bool = False,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return await receivable_service.list_receivables(
        db, customer_id=customer_id, status=status, open_only=open_only
    )


@router.post("/{receivable_id}/pay", response_model=ReceivableOut)
async def register_payment(
    receivable_id: uuid.UUID,
    data: PaymentCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return await receivable_service.register_payment(
        db, receivable_id, data.amount, current_user, data.notes
    )
